### PR TITLE
Allow "redeclaring" an initializer or method with non-overlapping availability.

### PR DESCRIPTION
**Pull request by jrose-apple**
This allows a library to "fix" an initializer or method that should have been failable or non-failable by gating the change on the user's version of Swift. Initializer failability and whether or not a method or initializer throws are two things that doesn't change the "overload signature", so normally the compiler would call this an invalid redeclaration. But since only one version can be active at a time based on the client context, there won't be a problem in practice.

[SR-4171](https://bugs.swift.org/browse/SR-4171) / rdar://problem/30470854

### CODE REVIEW

**Comment by jrose-apple** [comment]
I actually have even more tests I want to add but I can do that later.

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
I don't want to accidentally kill the long-tests I have running on the previous commit, but

@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test macOS

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -953,6 +953,61 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
...
+      // If the conflicting declarations have non-overlapping availability and
+      // - one throws and the other does not,
+      // - or they are initializers with different failability
+      // let them go.
+      bool isAcceptableVersionBasedChange = false;
+      {
+        auto *currentInit = dyn_cast<ConstructorDecl>(current);
+        auto *otherInit = dyn_cast<ConstructorDecl>(other);
+        if (currentInit && otherInit &&
+            ((currentInit->getFailability() == OTK_None) ^
```
Comment (jtbandes) [inline lib/Sema/TypeCheckDecl.cpp]: nit: `^` here and `!=` below; consider `!=` for both (IMO slightly clearer)

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -953,6 +953,61 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
           continue;
       }
 
+      // If the conflicting declarations have non-overlapping availability and
+      // - one throws and the other does not,
+      // - or they are initializers with different failability
+      // let them go.
+      bool isAcceptableVersionBasedChange = false;
+      {
+        auto *currentInit = dyn_cast<ConstructorDecl>(current);
```
Comment (jtbandes) [inline lib/Sema/TypeCheckDecl.cpp]: nit: `const`s?

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -953,6 +953,61 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
...
+        AvailabilityRange currentAvail = findSwiftAvailability(current);
+        AvailabilityRange otherAvail = findSwiftAvailability(other);
+        if (currentAvail.introduced && otherAvail.obsoleted &&
+            *currentAvail.introduced >= *otherAvail.obsoleted) {
+          continue;
+        }
+        if (otherAvail.introduced && currentAvail.obsoleted &&
+            *otherAvail.introduced >= *currentAvail.obsoleted) {
+          continue;
+        }
```
Comment (jtbandes) [inline lib/Sema/TypeCheckDecl.cpp]: minor suggestion: consider moving this logic into a helper, to the effect of `bool AvailabilityRange::precedes(const AvailabilityRange&) const` (there probably exist better naming choices, maybe even `operator<`)

**Review by jtbandes — COMMENTED** [review]
I'm not really qualified to review this as I don't fully understand the consequences of allowing such decls to survive `checkRedeclaration()` — but the code change looks fine to me 😃

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Interesting. Mentally I was considering this a disjunction of conditions and the below one a comparison of values, but if it's actually making it less clear then it's worth changing.

**Comment by jrose-apple** [comment]
There aren't too many people who've touched redeclaration checking at all; you're actually one of them. I'll wait a bit longer to give Doug a chance to look too, I guess.

**Comment by jtbandes** [comment]
I suppose I am, although my change was just disallowing some bad decls, whereas this allows decls to pass through that would previously have been rejected. I can't really evaluate what will happen downstream of that...

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Pull request merged**